src/cppext.cpp

```diff
@@ -313,6 +313,207 @@ cdd cdd_past(const cdd& state)
     return res;
 }
 
+/**
+ * Class for a 2D int32_t matrix where the number of rows is dynamic.
+ */
+class dynamic_two_dim_matrix
+{
+public:
+    /**
+     * Constructor for the \a dynamic_two_dim_matrix class.
+     * @param num_cols the fixed number of columns.
+     */
+    dynamic_two_dim_matrix(int num_cols)
+    {
+        assert(num_cols > 0);
+        this->num_cols = num_cols;
+
+        // Initialize the matrix.
+        matrix.emplace_back();
+        rows_to_be_ignored.push_back(false);
+    }
+
+    /**
+     * Add a new value at the end the current row as long as there is place left.
+     * @param value the value to add.
+     */
+    void add_value_to_row(int32_t value)
+    {
+        assert(matrix.back().size() < num_cols);
+        matrix.back().push_back(value);
+    }
+
+    /**
+     * Change the current row to the next row. Resizing is taking place when necessary. The values
+     * of the current row are copied to the next row up to and including the provided column.
+     * @param copy_to_column copy the previous values up to this column index (including this column).
+     */
+    void next_row(int copy_to_column)
+    {
+        assert(copy_to_column < num_cols);
+        assert(!matrix.empty());
+
+        std::vector<int32_t> new_row;
+
+        // Copy values from the previous row.
+        for (int j = 0; j <= copy_to_column; j++) {
+            new_row.push_back(matrix.at(matrix.size() - 1).at(j));
+        }
+        matrix.push_back(new_row);
+        rows_to_be_ignored.push_back(false);
+    }
+
+    /**
+     * Indicate that the current row needs to be ignored.
+     */
+    void ignore_current_row() { rows_to_be_ignored.back() = true; }
+
+    /**
+     * Delete the ignored rows from the matrix and shift the remaining rows upwards, i.e.,
+     * if row \a is to be ignored, the new row \i of the matrix is row \i + 1;
+     */
+    void delete_ignored_rows()
+    {
+        int i = 0;
+        for (auto it = matrix.begin(); it != matrix.end();) {
+            if (rows_to_be_ignored.at(i)) {
+                matrix.erase(it);
+            } else {
+                ++it;
+            }
+            ++i;
+        }
```

Comment:
Consider instead:
```suggestion
        matrix.erase(std::remove_if(matrix.begin(), matrix.end(), [i=0u](const auto& value) mutable {
            return rows_to_be_ignored.at(i++);
        }), matrix.end());
```

Reply:
Hmm.. It does not seem to be working correctly for some reason. The static test cases fail due to unexpected results in the `REQUIRE(...)` statements.

Reply:
I have updated the patch, please reconsider.